Tesla platform

zabuldon opened the pull request:
## Description:
Implementation of Tesla cars support

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#3273>

## Example entry for `configuration.yaml`:
```yaml
tesla:
  username: test@example.com
  password: mycoolpassword
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

zabuldon committed:
Tesla support implemetation

zabuldon committed:
requirements_all.txt fix

zabuldon committed:
.coveragerc fix

zabuldon committed:
logging-too-many-args fix

zabuldon committed:
logging-too-many-args attempt 2

dale3h commented on homeassistant/components/tesla.py:
```diff
@@ -0,0 +1,108 @@
+Support for Tesla cars.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/vera/
```
Please update the URL to https://home-assistant.io/components/tesla/

dale3h commented on homeassistant/components/tesla.py:
```diff
@@ -0,0 +1,108 @@
+
+    @property
+    def should_poll(self):
+        """Get polling requirement from vera device."""
```
Please correct the docstring.

dale3h reviewed (changes requested):
Preliminary review looks good aside from minor word changes. 🎃

MartinHjelmare commented on homeassistant/components/tesla.py:
```diff
@@ -0,0 +1,108 @@
+CONF_UPDATE_INTERVAL = 'update_interval'
+
+TESLA_ID_FORMAT = '{}_{}'
+TESLA_DEVICES = defaultdict(list)
```
Store this in the `hass.data` dict.
https://home-assistant.io/developers/code_review_component/#componentplatform-communication

MartinHjelmare commented on homeassistant/components/tesla.py:
```diff
@@ -0,0 +1,108 @@
+
+import voluptuous as vol
+
+from homeassistant.util import slugify
```
Please sort the import lines and imported names alphabetically, :abcd:, within each group builtin, 3rd party, home assistant.

MartinHjelmare commented on homeassistant/components/tesla.py:
```diff
@@ -0,0 +1,108 @@
+    email = config.get(CONF_EMAIL)
+    password = config.get(CONF_PASSWORD)
+    update_interval = config.get(CONF_UPDATE_INTERVAL)
+    TESLA_CONTROLLER = teslaApi(email, password, update_interval)
```
Store the tesla controller in `hass.data`.

dale3h commented on homeassistant/components/tesla.py:
```diff
@@ -0,0 +1,108 @@
+
+TESLA_CONTROLLER = None
+
+CONF_EMAIL = 'email'
```
Please use `username` instead of `email`.

MartinHjelmare commented on homeassistant/components/tesla.py:
```diff
@@ -0,0 +1,108 @@
+
+TESLA_CONTROLLER = None
+
+CONF_EMAIL = 'email'
```
Import this from const.py.

dale3h replied:
This should be changed to `username` as it is the standard when referring to user account credentials across all but 5 components/platforms.

dale3h commented on homeassistant/components/tesla.py:
```diff
@@ -0,0 +1,108 @@
+
+CONF_EMAIL = 'email'
+CONF_PASSWORD = 'password'
+CONF_UPDATE_INTERVAL = 'update_interval'
```
Please use `scan_interval` instead of `update_interval` (already declared in `homeassistant.const` as `CONF_SCAN_INTERVAL`).

MartinHjelmare commented on homeassistant/components/tesla.py:
```diff
@@ -0,0 +1,108 @@
+TESLA_CONTROLLER = None
+
+CONF_EMAIL = 'email'
+CONF_PASSWORD = 'password'
```
Import this from const.py.

dale3h replied:
Use `CONF_USERNAME`, `CONF_PASSWORD`, and `CONF_SCAN_INTERVAL` from `homeassistant.const`:

```python
from homeassistant.const import (
    CONF_USERNAME, CONF_PASSWORD, CONF_SCAN_INTERVAL)
```

MartinHjelmare commented on homeassistant/components/tesla.py:
```diff
@@ -0,0 +1,108 @@
+
+def setup(hass, base_config):
+    """Setup of Tesla platform."""
+    global TESLA_CONTROLLER
```
Remove.

dale3h commented on homeassistant/components/tesla.py:
```diff
@@ -0,0 +1,108 @@
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_EMAIL): cv.string,
```
Change to `CONF_USERNAME`.

dale3h commented on homeassistant/components/tesla.py:
```diff
@@ -0,0 +1,108 @@
+
+    config = base_config.get(DOMAIN)
+
+    email = config.get(CONF_EMAIL)
```
Change to `CONF_USERNAME`.

dale3h commented on homeassistant/components/tesla.py:
```diff
@@ -0,0 +1,108 @@
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_EMAIL): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_UPDATE_INTERVAL, default=300): cv.positive_int,
```
Change to `CONF_SCAN_INTERVAL`.

dale3h commented on homeassistant/components/tesla.py:
```diff
@@ -0,0 +1,108 @@
+
+    email = config.get(CONF_EMAIL)
+    password = config.get(CONF_PASSWORD)
+    update_interval = config.get(CONF_UPDATE_INTERVAL)
```
Change to `CONF_SCAN_INTERVAL`.

MartinHjelmare commented on homeassistant/components/tesla.py:
```diff
@@ -0,0 +1,108 @@
+    return True
+
+
+def map_tesla_device(tesla_device):
```
Why define a new function for this? I think you can just inline the property call above when appending devices in the defaultdict.

dale3h reviewed (dismissed):
* Use constants already provided in `homeassistant.const`
* Change all occurrences of `email` and `CONF_EMAIL` to `username` and `CONF_USERNAME`
* Change all occurrences of `update_interval` and `CONF_UPDATE_INTERVAL` to `scan_interval` and `CONF_SCAN_INTERVAL`

MartinHjelmare commented on homeassistant/components/tesla.py:
```diff
@@ -0,0 +1,108 @@
+
+        if self.tesla_device.has_battery():
+            attr[ATTR_BATTERY_LEVEL] = \
+                str(self.tesla_device.battery_level()) + '%'
```
Concatenating a unit to the value in the state attributes is not recommended and will make it harder for the user to eg use templates to manipulate state attributes of the device.

MartinHjelmare reviewed (changes requested):
Off to a very good start!

I've looked at the component so far.

balloob commented:
This is an awesome contribution @zabuldon ! 🎉

zabuldon committed:
Post-review fixes.

zabuldon committed:
requirements version fix

zabuldon committed:
requirements

zabuldon committed:
Lint fix

zabuldon commented:
I hope all issues were resolved.

zabuldon committed:
Hot fix

zabuldon committed:
:Merge branch 'dev' of https://github.com/zabuldon/home-assistant into dev

zabuldon committed:
requirements_all.txt fix

MartinHjelmare commented on homeassistant/components/climate/tesla.py:
```diff
@@ -0,0 +1,97 @@
+        return self._temperature
+
+    @property
+    def operation(self):
```
Remove this property. It's not used.

MartinHjelmare commented on homeassistant/components/climate/tesla.py:
```diff
@@ -0,0 +1,97 @@
+        TeslaDevice.__init__(self, tesla_device, controller)
+        self.entity_id = ENTITY_ID_FORMAT.format(self.tesla_id)
+        self._target_temperature = None
+        self._state = None
```
Remove this attribute. It won't be used.

MartinHjelmare commented on homeassistant/components/climate/tesla.py:
```diff
@@ -0,0 +1,97 @@
+        """Called by the Tesla device callback to update state."""
+        _LOGGER.debug('Updating: %s', self._name)
+        self.tesla_device.update()
+        self._state = self.tesla_device.is_hvac_enabled()
```
Remove.

MartinHjelmare commented on homeassistant/components/climate/tesla.py:
```diff
@@ -0,0 +1,97 @@
+    def temperature_unit(self):
+        """Return the unit of measurement."""
+        tesla_temp_units = (
+            self.tesla_device.measurement)
```
You don't need to break this line.

MartinHjelmare commented on homeassistant/components/climate/tesla.py:
```diff
@@ -0,0 +1,97 @@
+        _LOGGER.debug('Setting temperature for: %s', self._name)
+        temperature = kwargs.get(ATTR_TEMPERATURE)
+        if temperature:
+            self.tesla_device.set_temperature(kwargs.get(ATTR_TEMPERATURE))
```
```py
self.tesla_device.set_temperature(temperature)
```

MartinHjelmare reviewed (changes requested):
Looked at binary sensor and climate platforms.

zabuldon committed:
Review preparation.

zabuldon committed:
1. Linting fix.
2. Minimal value for SCAN_INTERVAL hardcoded to 300 sec (to prevent possible ban form Tesla)

zabuldon committed:
Removed redundant whitespace.

zabuldon committed:
Merge branch 'dev' of https://github.com/zabuldon/home-assistant into dev

MartinHjelmare commented on homeassistant/components/device_tracker/tesla.py:
```diff
@@ -0,0 +1,60 @@
+            tesla_id = slugify(device.uniq_name)
+            dev_id = slugify(name)
+
+            if dev_id is None:
```
`dev_id` can't be `None` after return from `slugify`. It will be a string. Otherwise it will be an error. If `name` can be `None` you need to check for that before passing it to `slugify`.

zabuldon replied:
It can't be None.
I've removed this part because dev_id is not used anywhere else. I didn't remember why I've added it...

MartinHjelmare commented on homeassistant/components/sensor/tesla.py:
```diff
@@ -0,0 +1,81 @@
+        super().__init__(tesla_device, controller)
+
+        if self.type:
+            self._name = self.tesla_device.name + ' ({})'.format(self.type)
```
```py
self._name = '{} ({})'.format(self.tesla_device.name, self.type)
```

MartinHjelmare commented on homeassistant/components/sensor/tesla.py:
```diff
@@ -0,0 +1,81 @@
+    @property
+    def unit_of_measurement(self):
+        """Return the unit_of_measurement of the device."""
+        return self.tesla_device.measurement
```
```py
return self._temperature_units
```

MartinHjelmare commented on homeassistant/components/device_tracker/tesla.py:
```diff
@@ -0,0 +1,60 @@
+class TeslaDeviceTracker(object):
+    """A class representing a Tesla device."""
+
+    def __init__(self, hass, config, see, tesla_devices, ):
```
Remove last comma and space.

MartinHjelmare reviewed (changes requested):
Looked at device_tracker, lock, and sensor.

zabuldon committed:
Fixed components according to @MartinHjelmare proposals and remarks.

MartinHjelmare commented on .coveragerc:
```diff
@@ -243,6 +243,7 @@ omit =
     homeassistant/components/binary_sensor/ping.py
     homeassistant/components/binary_sensor/rest.py
     homeassistant/components/binary_sensor/tapsaff.py
+    homeassistant/components/binary_sensor/tesla.py
```
Component with platforms have a group in the beginning of the file. Put an entry like this at the correct alphabetical place somewhere around line 170 where the t:s are:
```
homeassistant/components/tesla.py
homeassistant/components/*/tesla.py
```

MartinHjelmare reviewed (changes requested):
Last comment from me, I think.

Nice job!

zabuldon committed:
.coveragerc as @MartinHjelmare suggested.

fabaff committed:
Minor changes

fabaff committed:
Fix docstrings

fabaff committed:
Update ordering

fabaff committed:
Update quotes

fabaff committed:
Minor changes

fabaff committed:
Update quotes

dale3h commented:
Looks good! 🎃

dale3h commented:
I will make sure the Tesla docs are ready to go in the morning.
